### PR TITLE
Gh environments refactor

### PR DESCRIPTION
**Pull request by wesselj**
Propose migrating to use Github environments instead of secrets with naming convention.
Changes:
- Secrets are stored in Github environments (not applied to this repo yet).
- Configuration naming convention for functions and schedules follow environment instead of branch.
- Hierarchical configurations for functions:
  - Common for all functions per environment. Useful for client and tenant id. Is placed in common folder.
  - Common for function for all environments. General configuration that is not environment spesific.
  - Specific for function and envionment. For example data_set_id.
- Configurations removed from github action file other than environments and functions.

### CODE REVIEW

**Review by haakonvt — DISMISSED** [review]
I'm on parental leave for the forseeable future, so I would ask in #help-functions for a review, probably @qTipTip can be of help 😄 . I also know that @ddonukis has played around with - and made some very nice changes to this repo a lot lately, so I'm adding him as well.

File: .github/workflows/deploy-functions.yaml [hunk 1]
```diff
@@ -0,0 +1,105 @@
+---
+name: Deploy Functions to CDF project using OIDC
+on:
+  push:
+    branches: [main]
```
[the remark this answers is not in the document]
Comment (wesselj) [reply .github/workflows/deploy-functions.yaml:5]: I will change it to be master, though would it be a good idea to change from master to main, as I though we where sort of standardising on using 'main' branch?

File: README.md [hunk 2]
```diff
@@ -176,7 +190,7 @@ $ poetry run pre-commit run --all-files
 
 The template uses GitHub Actions to perform Code quality and Deployment operations. All workflows are located in the [.github/workflows](https://github.com/cognitedata/deploy-functions-oidc/tree/master/.github/workflows) folder.
 * `code-quality.yaml` Runs all code-quality checks and ensures the style, linting, basic static analysis
-* `deploy-push-master.yaml` Responsible for your PR deployment to customer environment(s)
+* `deploy-push-master.yaml` Responsible for you PR deployment to customer environment(s)
```
[the remark this answers is not in the document]
Comment (wesselj) [reply README.md:193]: ...and cloudy brain. Updating readme.

File: .github/workflows/deploy-functions.yaml [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        function: ${{ fromJson(needs.generate-matrix.outputs.folders) }}
+        # Add environments to deploy to.
+        # Each environment should have a corresponding github environment with secrets.
+        environment:
+          - test
+
+    environment: ${{ matrix.environment }}
+
+    steps:
+      - uses: actions/checkout@v3
```
[the remark this answers is not in the document]
Comment (wesselj) [reply .github/workflows/deploy-functions.yaml]: Ah, good catch, I updated README.

File: .github/workflows/deploy-functions.yaml [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+                echo "Config: $FILE not supplied!"
+            else
+                ALLKEYS=$(yq -o=tsv "keys" $FILE)
+                for CONFIGURATION in $ALLKEYS
+                do
+                    VALUE=$(yq -o=json -I=0 ".$CONFIGURATION" $FILE | sed -e 's/^"//' -e 's/"$//')
+                    echo $CONFIGURATION=$VALUE >> $GITHUB_OUTPUT
+                done
+            fi
+          done
```
[the remark this answers is not in the document]
Comment (wesselj) [reply .github/workflows/deploy-functions.yaml:72]: Initially if same key is defined in different config files, the key will be overridden by the most lokal to the given function.
So for test the hierarchy would be "common/function_config_test.yaml" -> "example_function1/function_config.yaml" -> "example_function1/function_config_test.yaml".
The reason for the multilevel config is that there are config that is common across functions on same environments such as tenant_id (it is possible that different environments span different tenants), soe configuration is common for a function across environments while some is specific to the function and environment such as client_id and data_set_id.

[the remark this answers is not in the document]
Comment (wesselj) [reply .github/workflows/deploy-functions.yaml:72]: So the somewhat confusing multi-structured config file with possibility to override configuration parameters is by design though I agree it does add some cognitive complexity.

**Pull request merged**